File: AST-Parser-main 2/test_samples/enterprise_crypto_service.py
"""
===============================================================================
Enterprise Cryptographic Microservice - AST Edge-Case Test Suite (Python)
===============================================================================
Comprehensive test suite containing PyCryptodome, cryptography.hazmat,
hashlib, aliased imports, multi-line decorators, and quantum edge cases.
"""

import os
import sys
import time
import json
import base64
from typing import Dict, Any, Optional, Tuple

# -----------------------------------------------------------------------------
# EDGE CASE 1: Aliased and Renamed Imports
# -----------------------------------------------------------------------------
from Crypto.Cipher import (
    AES as FastAESEngine,
    DES3 as TripleDESLegacyEngine,
    DES as SingleDESEngine,
    Blowfish as BlowfishLegacy,
    ARC4 as RC4StreamCipher,
    ChaCha20 as ChaCha20Modern
)
from Crypto.PublicKey import RSA as RSAPublicKeyManager, ECC as EllipticCurveManager
from Crypto.Random import get_random_bytes
from Crypto.Signature import pkcs1_15

import hashlib as GlobalHashEngine
from hashlib import (
    md5 as InsecureMD5Hash,
    sha1 as LegacySHA1Hash,
    sha256 as SecureSHA256Hash,
    sha512 as HighSecuritySHA512Hash
)

# -----------------------------------------------------------------------------
# EDGE CASE 2: Comments and String Literals (False Positive Traps)
# -----------------------------------------------------------------------------
# Developer Note: Never use MD5, DES3 or RSA-1024 in modern production!
# Reference: AES.MODE_ECB is dangerous because identical blocks yield identical ciphertext.
SYSTEM_LOG_TEMPLATE = "AUDIT_EVENT: Detected cipher initialization using AES-128 and DES"
INSECURE_ALGO_LIST = ["MD5", "SHA1", "DES", "3DES", "RC4"]


class SecurityAuditConfig:
    DEFAULT_DATA_LIFETIME_YEARS: int = 15
    PQC_MIGRATION_WINDOW_YEARS: int = 3
    CRQC_QUANTUM_ARRIVAL_YEARS: int = 10


# -----------------------------------------------------------------------------
# MODULE 1: Modern Quantum-Resistant Symmetric Handlers (AES-256 GCM / CBC)
# -----------------------------------------------------------------------------
class ModernEncryptionService:
    """Provides modern 256-bit symmetric encryption."""

    def __init__(self, key_256: Optional[bytes] = None):
        # 32 bytes = 256-bit key (Quantum Safe)
        self.key = key_256 or get_random_bytes(32)

    # EDGE CASE 3: Multi-line Call with named parameters
    def encrypt_aes_gcm(self, plaintext: bytes) -> Dict[str, bytes]:
        """Encrypts data using AES-256 in GCM authenticated mode."""
        cipher = FastAESEngine.new(
            self.key,
            FastAESEngine.MODE_GCM
        )
        ciphertext, tag = cipher.encrypt_and_digest(plaintext)
        return {
            "nonce": cipher.nonce,
            "ciphertext": ciphertext,
            "tag": tag,
            "algorithm": "AES-256-GCM",
            "security": "QUANTUM_SAFE"
        }

    def encrypt_aes_cbc(self, plaintext: bytes) -> Dict[str, bytes]:
        """Encrypts data using AES-256 in CBC mode with PKCS7 padding."""
        iv = get_random_bytes(16)
        cipher = FastAESEngine.new(
            self.key,
            FastAESEngine.MODE_CBC,
            iv=iv
        )
        pad_len = 16 - (len(plaintext) % 16)
        padded = plaintext + bytes([pad_len] * pad_len)
        ciphertext = cipher.encrypt(padded)
        return {
            "iv": iv,
            "ciphertext": ciphertext,
            "algorithm": "AES-256-CBC"
        }

    def encrypt_chacha20(self, plaintext: bytes) -> Dict[str, bytes]:
        """Encrypts data using modern ChaCha20 stream cipher."""
        cipher = ChaCha20Modern.new(key=self.key)
        ciphertext = cipher.encrypt(plaintext)
        return {
            "nonce": cipher.nonce,
            "ciphertext": ciphertext,
            "algorithm": "ChaCha20"
        }


# -----------------------------------------------------------------------------
# MODULE 2: Broken & Insecure Symmetric Ciphers (AES-ECB, 3DES, DES, Blowfish)
# -----------------------------------------------------------------------------
class InsecureLegacyCipherService:
    """Handles legacy decryption for backward compatibility (HIGH VULNERABILITY)."""

    def __init__(self):
        # 16 bytes = 128-bit key (Quantum Weakened by Grover)
        self.weak_key_128 = get_random_bytes(16)
        # 24 bytes = 3DES key
        self.des3_key = get_random_bytes(24)
        # 8 bytes = Single DES key (Classically Broken)
        self.des_key = get_random_bytes(8)
        # Blowfish key
        self.blowfish_key = get_random_bytes(16)

    # EDGE CASE 4: Dangerous ECB Mode Detection
    def insecure_aes_ecb(self, plaintext: bytes) -> bytes:
        """Insecure AES in Electronic Codebook Mode (Leaks data patterns)."""
        cipher = FastAESEngine.new(
            self.weak_key_128,
            FastAESEngine.MODE_ECB
        )
        pad_len = 16 - (len(plaintext) % 16)
        padded = plaintext + bytes([pad_len] * pad_len)
        return cipher.encrypt(padded)

    def insecure_3des_ecb(self, plaintext: bytes) -> bytes:
        """Insecure Triple-DES (Sweet32 vulnerability & Grover weakened)."""
        cipher = TripleDESLegacyEngine.new(
            self.des3_key,
            TripleDESLegacyEngine.MODE_ECB
        )
        pad_len = 8 - (len(plaintext) % 8)
        padded = plaintext + bytes([pad_len] * pad_len)
        return cipher.encrypt(padded)

    def insecure_single_des(self, plaintext: bytes) -> bytes:
        """Single 56-bit DES (Broken in 1999)."""
        cipher = SingleDESEngine.new(
            self.des_key,
            SingleDESEngine.MODE_CBC,
            iv=b"12345678"
        )
        pad_len = 8 - (len(plaintext) % 8)
        padded = plaintext + bytes([pad_len] * pad_len)
        return cipher.encrypt(padded)

    def legacy_blowfish(self, plaintext: bytes) -> bytes:
        """Legacy 64-bit Blowfish block cipher."""
        cipher = BlowfishLegacy.new(
            self.blowfish_key,
            BlowfishLegacy.MODE_CBC,
            iv=b"abcdefgh"
        )
        pad_len = 8 - (len(plaintext) % 8)
        padded = plaintext + bytes([pad_len] * pad_len)
        return cipher.encrypt(padded)

    def broken_rc4_stream(self, plaintext: bytes) -> bytes:
        """Broken RC4 / ARC4 stream cipher."""
        cipher = RC4StreamCipher.new(b"secret_key_12345")
        return cipher.encrypt(plaintext)


# -----------------------------------------------------------------------------
# MODULE 3: Quantum-Vulnerable Asymmetric Keys (RSA-1024, 2048, 4096, ECC)
# -----------------------------------------------------------------------------
class AsymmetricKeyManager:
    """Manages RSA and Elliptic Curve key pairs (Broken by Shor's Algorithm)."""

    @staticmethod
    def generate_rsa_key_pair(bits: int = 2048):
        """Generates RSA key pair (1024, 2048, 4096 bits)."""
        print(f"[*] Generating RSA Key with {bits} bits...")
        if bits == 1024:
            # Extreme Risk (Classically factorable)
            key = RSAPublicKeyManager.generate(1024)
        elif bits == 2048:
            # Standard RSA (Quantum Broken by Shor)
            key = RSAPublicKeyManager.generate(2048)
        elif bits == 4096:
            # High-Security RSA (Still Quantum Broken)
            key = RSAPublicKeyManager.generate(4096)
        else:
            key = RSAPublicKeyManager.generate(bits)

        private_pem = key.export_key()
        public_pem = key.publickey().export_key()
        return private_pem, public_pem

    @staticmethod
    def generate_ecc_curve():
        """Generates Elliptic Curve key on NIST P-256 curve (Broken by Shor)."""
        key = EllipticCurveManager.generate(curve="P-256")
        return key


# -----------------------------------------------------------------------------
# MODULE 4: Cryptographic Hashing Dispatcher (MD5, SHA1, SHA256, SHA512)
# -----------------------------------------------------------------------------
class IntegrityDigestService:
    """Computes cryptographic hashes across classical and quantum-safe standards."""

    @staticmethod
    def hash_md5_insecure(data: bytes) -> str:
        """Insecure MD5 hash (Collision vulnerable)."""
        h = InsecureMD5Hash()
        h.update(data)
        return h.hexdigest()

    @staticmethod
    def hash_sha1_legacy(data: bytes) -> str:
        """Insecure SHA-1 hash (SHAttered attack)."""
        h = LegacySHA1Hash()
        h.update(data)
        return h.hexdigest()

    @staticmethod
    def hash_sha256_secure(data: bytes) -> str:
        """Quantum-Safe SHA-256."""
        h = SecureSHA256Hash()
        h.update(data)
        return h.hexdigest()

    @staticmethod
    def hash_sha512_maximum(data: bytes) -> str:
        """High-entropy SHA-512."""
        h = HighSecuritySHA512Hash()
        h.update(data)
        return h.hexdigest()

    @staticmethod
    def generic_hashlib_calls(data: bytes):
        """Standard hashlib attribute calls."""
        d1 = GlobalHashEngine.md5(data).hexdigest()
        d2 = GlobalHashEngine.sha1(data).hexdigest()
        d3 = GlobalHashEngine.sha256(data).hexdigest()
        d4 = GlobalHashEngine.sha512(data).hexdigest()
        return d1, d2, d3, d4


# -----------------------------------------------------------------------------
# MODULE 5: Master Orchestrator Pipeline
# -----------------------------------------------------------------------------
class EnterpriseSecurityPipeline:
    """Main orchestration service running cryptographic workflows."""

    def __init__(self):
        self.modern = ModernEncryptionService()
        self.legacy = InsecureLegacyCipherService()
        self.asymmetric = AsymmetricKeyManager()
        self.digest = IntegrityDigestService()

    def run_full_suite(self):
        print("=" * 60)
        print("  Starting Enterprise Security Cryptographic Execution")
        print("=" * 60)

        test_payload = b"TopSecretEnterprisePayloadForSecurityAudit2026"

        # 1. Modern AES-256 & ChaCha20
        gcm_res = self.modern.encrypt_aes_gcm(test_payload)
        cbc_res = self.modern.encrypt_aes_cbc(test_payload)
        chacha_res = self.modern.encrypt_chacha20(test_payload)

        # 2. Insecure Legacy Ciphers
        ecb_res = self.legacy.insecure_aes_ecb(test_payload)
        des3_res = self.legacy.insecure_3des_ecb(test_payload)
        des_res = self.legacy.insecure_single_des(test_payload)
        blowfish_res = self.legacy.legacy_blowfish(test_payload)
        rc4_res = self.legacy.broken_rc4_stream(test_payload)

        # 3. Asymmetric Key Operations
        rsa_1024_priv, _ = self.asymmetric.generate_rsa_key_pair(1024)
        rsa_2048_priv, _ = self.asymmetric.generate_rsa_key_pair(2048)
        rsa_4096_priv, _ = self.asymmetric.generate_rsa_key_pair(4096)
        ecc_key = self.asymmetric.generate_ecc_curve()

        # 4. Hash Integrity Checksums
        h_md5 = self.digest.hash_md5_insecure(test_payload)
        h_sha1 = self.digest.hash_sha1_legacy(test_payload)
        h_sha256 = self.digest.hash_sha256_secure(test_payload)
        h_sha512 = self.digest.hash_sha512_maximum(test_payload)

        print("[+] All Cryptographic Workflows Initialized Successfully.")


if __name__ == "__main__":
    pipeline = EnterpriseSecurityPipeline()
    pipeline.run_full_suite()

File: AST-Parser-main 2/test_samples/enterprise_crypto_service.cpp
/**
 * ============================================================================
 * Enterprise Cryptographic Service - Edge-Case Test Suite (C++20)
 * ============================================================================
 * Contains realistic enterprise security routines, legacy authentication,
 * OpenSSL EVP wrappers, key exchange protocols, and AST parsing edge cases.
 */

#include <iostream>
#include <vector>
#include <string>
#include <memory>
#include <chrono>
#include <thread>
#include <map>

// OpenSSL Header Inclusions
#include <openssl/evp.h>
#include <openssl/rsa.h>
#include <openssl/pem.h>
#include <openssl/ec.h>
#include <openssl/aes.h>
#include <openssl/des.h>
#include <openssl/sha.h>
#include <openssl/md5.h>
#include <openssl/rand.h>

namespace enterprise::security {

// ----------------------------------------------------------------------------
// EDGE CASE 1: False Positive Traps in Comments & String Literals
// AST parsers MUST ignore these comments and plain strings!
// ----------------------------------------------------------------------------
// Insecure legacy note: Developers should avoid using MD5 and DES3 in production!
// Reference: RSA_generate_key_ex is deprecated in OpenSSL 3.0 in favor of EVP_PKEY.
const std::string AUDIT_LOG_HEADER = "CRITICAL_AUDIT: AES-128 and DES verification module initialized.";
const std::string MD5_WARNING_BANNER = "WARNING: System detected legacy MD5 signature from client request.";

enum class SecurityLevel {
    LEGACY_COMPATIBILITY,
    COMMERCIAL_STANDARD,
    POST_QUANTUM_READY
};

struct CipherSession {
    std::string sessionId;
    uint32_t keyLength;
    std::string algorithmName;
    std::vector<uint8_t> iv;
    std::vector<uint8_t> sessionKey;
    bool isAuthenticated;
};

// ----------------------------------------------------------------------------
// MODULE 1: Modern Quantum-Resistant Symmetric Encryption (AES-256-GCM)
// ----------------------------------------------------------------------------
class ModernDataEncryptor {
public:
    ModernDataEncryptor() : ctx_(EVP_CIPHER_CTX_new(), EVP_CIPHER_CTX_free) {}

    // EDGE CASE 2: Multi-line nested call expression
    bool encryptPayloadGCM(
        const std::vector<uint8_t>& plaintext,
        const std::vector<uint8_t>& key,
        const std::vector<uint8_t>& iv,
        std::vector<uint8_t>& ciphertext,
        std::vector<uint8_t>& tag
    ) {
        if (key.size() != 32 || iv.size() != 12) {
            return false;
        }

        int len = 0;
        int ciphertext_len = 0;
        ciphertext.resize(plaintext.size());
        tag.resize(16);

        // AST Target: EVP_aes_256_gcm()
        if (1 != EVP_EncryptInit_ex(
                     ctx_.get(),
                     EVP_aes_256_gcm(),
                     nullptr,
                     nullptr,
                     nullptr
                 )) {
            return false;
        }

        if (1 != EVP_CIPHER_CTX_ctrl(ctx_.get(), EVP_CTRL_GCM_SET_IVLEN, static_cast<int>(iv.size()), nullptr)) {
            return false;
        }

        if (1 != EVP_EncryptInit_ex(ctx_.get(), nullptr, nullptr, key.data(), iv.data())) {
            return false;
        }

        if (1 != EVP_EncryptUpdate(ctx_.get(), ciphertext.data(), &len, plaintext.data(), static_cast<int>(plaintext.size()))) {
            return false;
        }
        ciphertext_len = len;

        if (1 != EVP_EncryptFinal_ex(ctx_.get(), ciphertext.data() + len, &len)) {
            return false;
        }
        ciphertext_len += len;

        EVP_CIPHER_CTX_ctrl(ctx_.get(), EVP_CTRL_GCM_GET_TAG, 16, tag.data());
        ciphertext.resize(ciphertext_len);
        return true;
    }

    // AST Target: ChaCha20-Poly1305 Modern Stream Cipher
    bool encryptChaCha20(
        const std::vector<uint8_t>& plaintext,
        const std::vector<uint8_t>& key,
        const std::vector<uint8_t>& nonce,
        std::vector<uint8_t>& ciphertext
    ) {
        int len = 0;
        ciphertext.resize(plaintext.size() + 16);

        if (1 != EVP_EncryptInit_ex(
                     ctx_.get(),
                     EVP_chacha20_poly1305(),
                     nullptr,
                     key.data(),
                     nonce.data()
                 )) {
            return false;
        }

        EVP_EncryptUpdate(ctx_.get(), ciphertext.data(), &len, plaintext.data(), static_cast<int>(plaintext.size()));
        EVP_EncryptFinal_ex(ctx_.get(), ciphertext.data() + len, &len);
        return true;
    }

private:
    std::unique_ptr<EVP_CIPHER_CTX, decltype(&EVP_CIPHER_CTX_free)> ctx_;
};

// ----------------------------------------------------------------------------
// MODULE 2: Legacy Quantum-Vulnerable Asymmetric Key Generation (RSA-1024, 2048, 4096)
// ----------------------------------------------------------------------------
class LegacyKeyManager {
public:
    // EDGE CASE 3: Parameter inspection for variable key sizes (1024 vs 2048 vs 4096)
    static bool generateLegacyRSAKey(int keyBits) {
        std::cout << "[*] Initializing RSA Key Generation for bits: " << keyBits << "\n";
        
        BIGNUM* bn = BN_new();
        BN_set_word(bn, RSA_F4);

        RSA* rsa = RSA_new();
        
        // AST Target: RSA_generate_key_ex with dynamic bits parameter
        if (keyBits == 1024) {
            // High vulnerability legacy key
            RSA_generate_key_ex(rsa, 1024, bn, nullptr);
        } else if (keyBits == 2048) {
            // Standard commercial RSA key (Quantum Broken)
            RSA_generate_key_ex(rsa, 2048, bn, nullptr);
        } else if (keyBits == 4096) {
            // Extended RSA key (Still Quantum Broken by Shor)
            RSA_generate_key_ex(rsa, 4096, bn, nullptr);
        }

        RSA_free(rsa);
        BN_free(bn);
        return true;
    }

    // AST Target: Elliptic Curve Key Generation (ECC Curve P-256)
    static bool generateEllipticCurveKey() {
        // Quantum Vulnerable ECC (Broken by Shor)
        EC_KEY* ecKey = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
        if (ecKey) {
            EC_KEY_generate_key(ecKey);
            EC_KEY_free(ecKey);
            return true;
        }
        return false;
    }
};

// ----------------------------------------------------------------------------
// MODULE 3: Weak & Broken Legacy Ciphers (3DES-CBC, DES-ECB, AES-128-CBC)
// ----------------------------------------------------------------------------
class LegacySymmetricFallback {
public:
    // AST Target: Insecure 3DES Block Cipher
    static void legacy3DESEncryption(const uint8_t* in, uint8_t* out, size_t length, const uint8_t* key) {
        EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
        
        // Insecure 3DES call (Grover weakened & Sweet32 birthday collision attack)
        EVP_EncryptInit_ex(
            ctx,
            EVP_des_ede3_cbc(),
            nullptr,
            key,
            nullptr
        );
        
        int outlen = 0;
        EVP_EncryptUpdate(ctx, out, &outlen, in, static_cast<int>(length));
        EVP_EncryptFinal_ex(ctx, out + outlen, &outlen);
        EVP_CIPHER_CTX_free(ctx);
    }

    // AST Target: Weak AES-128-CBC (Quantum Weakened)
    static void weakAES128CBC(const uint8_t* input, uint8_t* output, int len, const uint8_t* k, const uint8_t* iv) {
        EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
        
        EVP_EncryptInit_ex(
            ctx,
            EVP_aes_128_cbc(),
            nullptr,
            k,
            iv
        );

        int outlen = 0;
        EVP_EncryptUpdate(ctx, output, &outlen, input, len);
        EVP_EncryptFinal_ex(ctx, output + outlen, &outlen);
        EVP_CIPHER_CTX_free(ctx);
    }

    // AST Target: Low-level AES_set_encrypt_key
    static void lowLevelAESRaw(const unsigned char* userKey, AES_KEY* key) {
        AES_set_encrypt_key(userKey, 128, key);
    }
};

// ----------------------------------------------------------------------------
// MODULE 4: Cryptographic Hashing Engine (MD5, SHA-1, SHA-256, SHA-512)
// ----------------------------------------------------------------------------
class IntegrityVerifier {
public:
    // AST Target: Insecure MD5 Hash
    static void computeMD5(const std::string& input, unsigned char* digest) {
        // Broken by classical collision attacks
        EVP_MD_CTX* mdctx = EVP_MD_CTX_new();
        EVP_DigestInit_ex(mdctx, EVP_md5(), nullptr);
        EVP_DigestUpdate(mdctx, input.c_str(), input.length());
        EVP_DigestFinal_ex(mdctx, digest, nullptr);
        EVP_MD_CTX_free(mdctx);
    }

    // AST Target: Insecure SHA-1 Hash
    static void computeSHA1(const std::string& input, unsigned char* digest) {
        // Broken by SHAttered collision attack
        EVP_MD_CTX* mdctx = EVP_MD_CTX_new();
        EVP_DigestInit_ex(mdctx, EVP_sha1(), nullptr);
        EVP_DigestUpdate(mdctx, input.c_str(), input.length());
        EVP_DigestFinal_ex(mdctx, digest, nullptr);
        EVP_MD_CTX_free(mdctx);
    }

    // AST Target: Secure SHA-256 (Quantum Resistant)
    static void computeSHA256(const std::string& input, unsigned char* digest) {
        EVP_MD_CTX* mdctx = EVP_MD_CTX_new();
        EVP_DigestInit_ex(mdctx, EVP_sha256(), nullptr);
        EVP_DigestUpdate(mdctx, input.c_str(), input.length());
        EVP_DigestFinal_ex(mdctx, digest, nullptr);
        EVP_MD_CTX_free(mdctx);
    }

    // AST Target: High-Security SHA-512
    static void computeSHA512(const std::string& input, unsigned char* digest) {
        EVP_MD_CTX* mdctx = EVP_MD_CTX_new();
        EVP_DigestInit_ex(mdctx, EVP_sha512(), nullptr);
        EVP_DigestUpdate(mdctx, input.c_str(), input.length());
        EVP_DigestFinal_ex(mdctx, digest, nullptr);
        EVP_MD_CTX_free(mdctx);
    }
};

// ----------------------------------------------------------------------------
// MODULE 5: Master Orchestration Dispatcher
// ----------------------------------------------------------------------------
class EnterpriseCryptoService {
public:
    EnterpriseCryptoService() : encryptor_(std::make_unique<ModernDataEncryptor>()) {}

    void executeFullSecurityAudit() {
        std::cout << "[+] Running Enterprise Cryptographic Security Audit...\n";

        // Trigger RSA routines
        LegacyKeyManager::generateLegacyRSAKey(1024);
        LegacyKeyManager::generateLegacyRSAKey(2048);
        LegacyKeyManager::generateLegacyRSAKey(4096);
        LegacyKeyManager::generateEllipticCurveKey();

        // Trigger Hash routines
        unsigned char md5_buf[16];
        unsigned char sha1_buf[20];
        unsigned char sha256_buf[32];
        unsigned char sha512_buf[64];
        IntegrityVerifier::computeMD5("secret_auth_token", md5_buf);
        IntegrityVerifier::computeSHA1("session_handshake", sha1_buf);
        IntegrityVerifier::computeSHA256("secure_transaction", sha256_buf);
        IntegrityVerifier::computeSHA512("admin_master_credential", sha512_buf);

        // Trigger AES and 3DES routines
        std::vector<uint8_t> payload = {0xDE, 0xAD, 0xBE, 0xEF, 0x01, 0x02, 0x03, 0x04};
        std::vector<uint8_t> key256(32, 0x42);
        std::vector<uint8_t> iv12(12, 0x01);
        std::vector<uint8_t> ciphertext;
        std::vector<uint8_t> tag;

        encryptor_->encryptPayloadGCM(payload, key256, iv12, ciphertext, tag);

        uint8_t des_out[64];
        uint8_t des_key[24] = {0};
        LegacySymmetricFallback::legacy3DESEncryption(payload.data(), des_out, payload.size(), des_key);

        std::cout << "[+] Enterprise Cryptographic Audit Completed.\n";
    }

private:
    std::unique_ptr<ModernDataEncryptor> encryptor_;
};

} // namespace enterprise::security

// Entry Point Test Wrapper
int main() {
    enterprise::security::EnterpriseCryptoService service;
    service.executeFullSecurityAudit();
    return 0;
}
